src/ripple/basics/feeunits.h

```diff
@@ -0,0 +1,540 @@
+    TaggedFee (value_type fee)
+        : fee_ (fee)
+    {
+    }
+
+    template <class Other,
+        class = enable_if_compatible_t <Other>>
+    constexpr
+    explicit
+    TaggedFee (Other fee)
+        : fee_ (static_cast<value_type> (fee))
+    {
+    }
+
+    template <class OtherFee,
+        class = enable_if_compatiblefee_t<OtherFee> >
+    constexpr
+    TaggedFee (OtherFee const& fee)
+        : fee_ (static_cast<value_type> (fee.value()))
+    {
+    }
+
+    template <class Other,
+        class = enable_if_compatible_t <Other>>
+    TaggedFee&
+    operator= (Other fee)
+    {
+        fee_ = static_cast<value_type> (fee);
+        return *this;
+    }
+
+    template <class OtherFee,
+        class = enable_if_compatiblefee_t <OtherFee>>
+    TaggedFee&
+    operator= (OtherFee const& fee)
+    {
+        fee_ = static_cast<value_type> (fee.value());
+        return *this;
+    }
+
+    constexpr
+    TaggedFee
+    operator* (value_type const& rhs) const
+    {
+        return TaggedFee{ fee_ * rhs };
+    }
+
+    constexpr
+    value_type
+    operator/ (TaggedFee const& rhs) const
+    {
+        return fee_ / rhs.fee_;
+    }
+
+    TaggedFee&
+    operator+= (TaggedFee const& other)
+    {
+        fee_ += other.fee();
+        return *this;
+    }
+
+    TaggedFee&
+    operator-= (TaggedFee const& other)
+    {
+        fee_ -= other.fee();
+        return *this;
+    }
+
+    TaggedFee&
+    operator++()
+    {
+        ++fee_;
+        return *this;
+    }
+
+    TaggedFee&
+    operator--()
+    {
+        --fee_;
+        return *this;
+    }
+
+    TaggedFee&
+    operator*= (value_type const& rhs)
+    {
+        fee_ *= rhs;
+        return *this;
+    }
+
+    TaggedFee&
+    operator/= (value_type const& rhs)
+    {
+        fee_ /= rhs;
+        return *this;
+    }
+
+    TaggedFee&
+    operator%= (value_type const& rhs)
+    {
+        fee_ %= rhs;
+        return *this;
+    }
+
+    TaggedFee
+    operator- () const
+    {
+        static_assert( std::is_unsigned_v<T>,
+            "- operator illegal on unsigned fee types");
+        return { -fee_ };
+    }
+
+    bool
+    operator==(TaggedFee const& other) const
+    {
+        return fee_ == other.fee_;
+    }
+
+    bool
+    operator<(TaggedFee const& other) const
+    {
+        return fee_ < other.fee_;
+    }
+
+    /** Returns true if the amount is not zero */
+    explicit
+    constexpr
+    operator bool() const noexcept
+    {
+        return fee_ != 0;
+    }
+
+    /** Return the sign of the amount */
+    constexpr
+    int
+    signum() const noexcept
+    {
+        return (fee_ < 0) ? -1 : (fee_ ? 1 : 0);
+    }
+
+    /** Returns the number of drops */
+    constexpr
+    value_type
+    fee () const
+    {
+        return fee_;
+    }
+
+    Json::Value
+    json () const
+    {
+        static_assert(is_usable_unit_v<TaggedFee>, "Invalid unit type");
+        using jsontype = std::conditional_t<std::is_signed_v<value_type>,
```

Comment:
This won't work for floating point types. Should we restrict this class to integer types? (I do propose getting rid of the one use of `double` for this class in another comment).

Reply:
Yes, I think it makes sense to restrict this function to integer types. But I figured out how to overload depending on the `value_type`, so it'll work for both the integer and floating types. I'm never 100% confident with template metaprogramming, so I'm all ears if there's a better way to implement it.

Reply:
When 1.4 b1lands we'll get `if constexpr`. I'd probably use that over `enable_if` here. I'm also not sure why we need the `is_usable_unit` check.

Reply:
Oh yeah! That'll be much nicer!

> I'm also not sure why we need the `is_usable_unit` check.

Future-proofing. I should document that. Right now, all defined unit tags are valid, but if this class is used in a more general sense later (and I can imagine it being useful for anything that needs a compile-time unit), then I don't want new units to be convertible to JSON by default. There was an example with that `xrp_unit_product_tag` tag before I got rid of it at your very helpful suggestion. They can be added to that value later if it's appropriate, but I want it to be intentional.